Firestore/core/src/firebase/firestore/util/async_queue.cc

```diff
@@ -127,9 +146,16 @@ AsyncQueue::Operation AsyncQueue::Wrap(const Operation& operation) {
   // Decorator pattern: wrap `operation` into a call to `ExecuteBlocking` to
   // ensure that it doesn't spawn any nested operations.
 
-  // Note: can't move `operation` into lambda until C++14.
-  auto shared_this = shared_from_this();
-  return [shared_this, operation] { shared_this->ExecuteBlocking(operation); };
+  // Capture by weak_ptr in order to avoid implicitly keeping the AsyncQueue
+  // alive even though the owning FirestoreClient may have been destroyed
+  // already.
+  std::weak_ptr<AsyncQueue> weak_this = shared_from_this();
+  return [weak_this, operation] {
+    std::shared_ptr<AsyncQueue> shared_this = weak_this.lock();
```

Comment:
Optional: use `auto`? (the type seems obvious enough, given that the `weak_ptr` declaration is just two lines above) No strong feelings, though.